File: leetcode/1-100/42-trapping-rain-water/bench/trapping_rain_water.py
"""LeetCode #42 bench mirror — Python, the converging two-pointer solver (★).

Mirrors bench/trapping_rain_water.kara: advance the shorter outer wall, settling each column
with its running max. The buffer is reused in place each iteration with a k-dependent jagged
terrain; fold each answer into a rolling checksum. Same workload + sink as every other
mirror. This is the slow interpreted baseline (long-workload lane).
"""


def trap(height: list[int], n: int) -> int:
    left, right = 0, n - 1
    left_max = right_max = 0
    water = 0
    while left < right:
        if height[left] < height[right]:
            if height[left] >= left_max:
                left_max = height[left]
            else:
                water += left_max - height[left]
            left += 1
        else:
            if height[right] >= right_max:
                right_max = height[right]
            else:
                water += right_max - height[right]
            right -= 1
    return water


def main() -> None:
    total = 200000
    n = 1000
    modulus = 1000000007

    height = [(i * 37) % 100 for i in range(n)]
    acc = 0
    for k in range(total):
        height[k % n] = (k * 19) % 100
        ans = trap(height, n)
        acc = (acc * 131 + ans) % modulus

    print(acc)


if __name__ == "__main__":
    main()

File: leetcode/1-100/42-trapping-rain-water/bench/trapping_rain_water.rs
// LeetCode #42 bench mirror — Rust, the converging two-pointer solver (★).
//
// Mirrors bench/trapping_rain_water.kara exactly: advance the shorter outer wall, settling
// each column with its running max. The buffer is allocated once and overwritten in place
// each iteration with a k-dependent jagged terrain; fold each answer into a rolling checksum.
// Same workload + sink as every other mirror.

fn trap(height: &[i64], n: i64) -> i64 {
    let mut left = 0i64;
    let mut right = n - 1;
    let mut left_max = 0i64;
    let mut right_max = 0i64;
    let mut water = 0i64;
    while left < right {
        if height[left as usize] < height[right as usize] {
            if height[left as usize] >= left_max {
                left_max = height[left as usize];
            } else {
                water += left_max - height[left as usize];
            }
            left += 1;
        } else {
            if height[right as usize] >= right_max {
                right_max = height[right as usize];
            } else {
                water += right_max - height[right as usize];
            }
            right -= 1;
        }
    }
    water
}

fn main() {
    let total: i64 = 200000;
    let n: i64 = 1000;
    let modulus: i64 = 1000000007;

    let mut height: Vec<i64> = vec![0; n as usize];
    let mut i = 0i64;
    while i < n {
        height[i as usize] = (i * 37) % 100;
        i += 1;
    }

    let mut acc: i64 = 0;
    let mut k: i64 = 0;
    while k < total {
        height[(k % n) as usize] = (k * 19) % 100;
        let ans = trap(&height, n);
        acc = (acc * 131 + ans) % modulus;
        k += 1;
    }

    println!("{}", acc);
}
